### ISSUE TITLE
object detection learn background

### ISSUE THREAD

**Issue by HuangBo-Terraloupe** [yhenon/keras-frcnn#228]
I want to ask a general question about the detection, hope can get some ideas.

Now I am doing a project to detection a object in a wide range. To train the faster rcnn, I only use the image patches that contains at least one object. Faster rcnn will take the Region Proposal, whose IOU with any groundtruth bounding box is less than 30% as negative samples. But for the background that far away from the object, the network never see them. So the network somehow think for every image, at least there is a object inside it. (of course we can also play around with number of bbox_threshold=0.8)
 
My question is how can I add other background as negative samples into the training set. So that the faster rcnn can learn from it.

I tried put these background image into training set, it seems will raise a error and pass that image.

**Comment by HuangBo-Terraloupe** [comment]
@yhenon @Abhijit-2592
Do you have some idea how to solve these problem?
For the background image, can I just make a fake bounding box like (0, 0, 1, 1), and then sign a exist class to it, so that I can put these images as negative samples into training set.

**Comment by Abhijit-2592** [comment]
@HuangBo-Terraloupe   sorry for the late reply. I don't seem to understand your question. Could you upload an image or something?

**Comment by HuangBo-Terraloupe** [comment]
@Abhijit-2592 

I want to detect some objects (buildings) in a big map, for example the map of a country. 
Let's say the map size is 10000*10000. For training, you need crop the map into image patches, like 500*500. So some image patch contains building, some image patch are just background, do not contain any building.
My question is, how to add these background images as negative samples for training the network. So later if I use 'sliding window' go through the whole map, the network will also perform good for these background images.
I mean for these background images, my detection network should not generate any bounding box, for the image that contains a building, my network should detect it.

**Comment by Abhijit-2592** [comment]
@HuangBo-Terraloupe . Yeah Now I understand what you want. It is technically called as hard negative mining. I don't have practical experience with it though but, this is what I can think off from top of my head.

Say you have a big image of size 10k x 10k pixels. You are cropping out regions of size 500x500 pixels from it. It is called as sliding window approach which is a method to detect small objects in an image. For simplicity lets just consider 2 (A,B) 500x500 images cropped out of the 10k x 10k image and assume in image A you have a object(a building in your case)of size 200 x 200 and in image B you don't have any object. Lets also assume that you have anchor boxes of sizes [128,256,512] with the corresponding aspect ratios. In image A you have a positive example. From Image B you can draw a bounding box of arbitrary size but comparable to the anchor box scales and label it as 'bg' (specifically) and the simple_parser.py script in this repo should take care of it.

But if you want more control you can change the ratio of positive and negative samples that goes inside the model_classifier() in train.py script. By default it is set it to approximately 1.

Let me know if it works
Cheers
